File: lclang/src/main/java/lclang/LCRootExecutor.java
package lclang;

import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.LCClass;
import lclang.libs.lang.classes.NullClass;
import lclang.statements.Statement;

import java.util.ArrayList;
import java.util.HashMap;

public class LCRootExecutor {
    public final String path;
    public final LCBaseExecutor executor = new LCBaseExecutor(this);
    public final HashMap<String, LCClass> globals = new HashMap<>();
    public final HashMap<String, LCClass> classes = new HashMap<>();
    public final ArrayList<Statement> statements = new ArrayList<>();

    public LCRootExecutor(String path) {
        this.path = path;
    }

    public void accept(LCRootExecutor root) {
        classes.putAll(root.classes);
        globals.putAll(root.globals);
    }

    public Value getLink(Caller caller, String name) throws LCLangRuntimeException {
        LCClass clazz = globals.getOrDefault(name, null);
        if(clazz==null) return null;

        return clazz.asValue();
    }

    public LCClass execute() throws LCLangRuntimeException {
        for (Statement statement: statements) {
            Caller caller = statement.getCaller(this);
            Value value = statement.visit(caller, executor);

            LCClass clazz = value.get.invoke(caller);
            if(value.state!=Value.State.NOTHING)
                return clazz;
        }

        return NullClass.instance;
    }
}

File: lclang/src/main/kotlin/lclang/LCRootExecutor.kt
package lclang

import lclang.statements.Statement

open class LCRootExecutor(
    val path: String
) {
    val executor = LCBaseExecutor(this)
    val globals = HashMap<String, Value>()
    val classes = HashMap<String, LCClass>()
    val statements = ArrayList<Statement>()

    fun accept(root: LCRootExecutor){
        classes.putAll(root.classes)
        globals.putAll(root.globals)
    }

    fun execute(): LCClass? {
//        val classes = ctx.classExpr()
//        for(clazz in classes)
//            this.classes[clazz.name.text] = LCClass.from("",this, clazz)
//
//        val components = ctx.component()
//        for(component in components){
//            val name = component.type().namedType().text + "\\"
//
//            val componentGlobals = component.global()
//            for(global in componentGlobals)
//                globals[name+global.ID().text] = compiler.visitValue(global.value())
//                    .visit(Caller(this, 0, 0), executor)
//
//            val componentClasses = component.classExpr()
//            for(clazz in componentClasses)
//                this.classes[name+clazz.name] = LCClass.from(name,this, clazz)
//        }
//
//        val files = HashMap<String, LCRootExecutor>()
//        for(usage in ctx.use()){
//            val requiredFile = File(File(path).parent, usage.STRING().text
//                .substring(1).substringBeforeLast('"'))
//
//            var eval: LCRootExecutor? = null
//            if(!requiredFile.exists())
//                throw LCLangException("File", "file "+requiredFile.name+" not found",
//                    Caller(this, usage.start.line, usage.stop.line))
//            else if(files[requiredFile.path]!=null)
//                eval = files[requiredFile.path]!!
//            else if(requiredFile.length()!=0L){
//                eval = LCRootExecutor(requiredFile.path.toString())
//                eval.runInput(requiredFile.readText())
//
//                files[requiredFile.path] = eval
//            }
//
//            if(eval!=null){
//                val name = usage.namedType().text
//                if(usage.useGlobal!=null){
//                    if(eval.globals.containsKey(name))
//                        globals[name] = eval.globals[name]!!
//                    else throw LCLangException("File", "global "+requiredFile.name+" not found",
//                        Caller(this, usage.start.line, usage.stop.line))
//                }else{
//                    if(eval.classes.containsKey(name))
//                        this.classes[name] = eval.classes[name]!!
//                    else throw LCLangException("File", "class "+requiredFile.name+" not found",
//                        Caller(this, usage.start.line, usage.stop.line))
//                }
//            }
//        }
//
//        files.clear()
//
//        for(global in ctx.global())
//            globals[global.ID().text] = compiler.visitValue(global.value())
//                .visit(Caller(this, 0, 0), executor)
//
//        for(method in ctx.method()) {
//            globals[method.ID().text] = ContextMethod(this.executor, method).asValue()
//        }

        val rootCaller = Caller(this, 0, 0)
        for(stmt in statements) {
            val caller = stmt.getCaller(rootCaller)
            val value = stmt.visit(caller, executor)

            val clazz = value.get(caller)
            if(value.state!=Value.State.NOTHING)
                return clazz
        }

        return null
    }
}